sample-parent/sample-features/checkpoint-runner/src/main/java/org/talend/sdk/component/sample/feature/checkpoint/CheckpointInput.java

```diff
@@ -0,0 +1,207 @@
+
+import javax.annotation.PostConstruct;
+import javax.json.JsonBuilderFactory;
+import javax.json.JsonObject;
+
+import org.talend.sdk.component.api.component.Icon;
+import org.talend.sdk.component.api.component.MigrationHandler;
+import org.talend.sdk.component.api.component.Version;
+import org.talend.sdk.component.api.configuration.Option;
+import org.talend.sdk.component.api.configuration.type.DataSet;
+import org.talend.sdk.component.api.configuration.type.DataStore;
+import org.talend.sdk.component.api.configuration.ui.DefaultValue;
+import org.talend.sdk.component.api.configuration.ui.layout.GridLayout;
+import org.talend.sdk.component.api.input.Emitter;
+import org.talend.sdk.component.api.input.Producer;
+import org.talend.sdk.component.api.input.checkpoint.Checkpoint;
+import org.talend.sdk.component.api.input.checkpoint.CheckpointAvailable;
+import org.talend.sdk.component.api.input.checkpoint.CheckpointData;
+import org.talend.sdk.component.api.meta.Documentation;
+
+import lombok.Data;
+
+@Version
+@Icon(value = Icon.IconType.CUSTOM, custom = "icon")
+@Documentation("Checkpoint Input sample processor connector.")
+@Emitter(family = "checkpoint", name = "input")
+public class CheckpointInput implements Serializable {
+
+    private final transient JsonBuilderFactory factory;
+
+    private List<Integer> data;
+
+    private transient ListIterator<Integer> iterator;
+
+    private Integer bookmark;
+
+    private boolean newBookmark;
+
+    private final transient InputConfig configuration;
+
+    public CheckpointInput(final JsonBuilderFactory factory, @Option("configuration") final InputConfig config) {
+        this.factory = factory;
+        this.configuration = config;
+    }
+
+    @PostConstruct
+    public void init() {
+        data = IntStream.range(0, configuration.dataset.maxRecords).boxed().collect(toList());
+        if (configuration.checkpoint == null) {
+            bookmark = 0;
+        } else {
+            if ("finished".equals(configuration.checkpoint.status)) {
+                bookmark = 0;
+            } else {
+                if (configuration.checkpoint.sinceId > configuration.dataset.maxRecords) {
+                    bookmark = 0;
+                } else {
+                    bookmark = configuration.checkpoint.sinceId;
+                }
+            }
+        }
+        if (bookmark == null) {
+            iterator = new ArrayList<Integer>().listIterator();
+        } else {
+            iterator = data.listIterator(bookmark);
+        }
+    }
+
+    @Producer
+    public JsonObject data() {
+        if (iterator == null) {
+            iterator = data.listIterator();
+        }
+        final Integer produced = iterator.hasNext() ? iterator.next() : null;
+        if (produced == null) {
+            configuration.checkpoint.status = "finished";
+            configuration.checkpoint.sinceId = bookmark;
+            return null;
+        }
+        bookmark = produced;
```

Comment:
maybe it's better to change the example input it's a bit confusing that bookmark kind of iniialized by data. What is bookmark? Is it some kind of checkpoint? can it be an incrementation? bookmark++

Reply:
here, the bookmark match the latest checkpoint passed.